tests/MongoDB.Driver.Core.Tests/Core/Authentication/ScramSha256AuthenticatorTests.cs

```diff
@@ -0,0 +1,265 @@
+﻿/* Copyright 2018-present MongoDB Inc.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System;
+using System.Net;
+using System.Threading;
+using FluentAssertions;
+using MongoDB.Bson;
+using MongoDB.Driver.Core.Authentication;
+using MongoDB.Driver.Core.Clusters;
+using MongoDB.Driver.Core.Servers;
+using MongoDB.Driver.Core.Helpers;
+using MongoDB.Driver.Core.Misc;
+using MongoDB.Driver.Core.WireProtocol.Messages;
+using Xunit;
+using MongoDB.Driver.Core.Connections;
+using MongoDB.Bson.TestHelpers.XunitExtensions;
+namespace MongoDB.Driver.Core.Authentication
+{
+    public class ScramSha256AuthenticatorTests
+    {
+        private static readonly UsernamePasswordCredential __credential
+            = new UsernamePasswordCredential("source", "user", "pencil");
+        private static readonly ClusterId __clusterId = new ClusterId();
+        private static readonly ServerId __serverId = new ServerId(__clusterId, new DnsEndPoint("localhost", 27017));
+        private static readonly ConnectionDescription __description = new ConnectionDescription(
+            new ConnectionId(__serverId),
+            new IsMasterResult(new BsonDocument("ok", 1).Add("ismaster", 1)),
+            new BuildInfoResult(new BsonDocument("version", "3.7.3")));
```

Comment:
Technically requires 4.0 compatibility mode to be enabled, so I'll update this to 4.0.0